pkg/dependencies/factory.go

```diff
@@ -504,20 +504,47 @@ func (f *Factory) WithGovc() *Factory {
 	return f
 }
 
-func (f *Factory) WithCmk() *Factory {
+// WithCmkFromEnv initializes the CMK executable for the object being constructed to make it available in the constructor.
+func (f *Factory) WithCmkFromEnv() *Factory {
 	f.WithExecutableBuilder().WithWriter()
 
 	f.buildSteps = append(f.buildSteps, func(ctx context.Context) error {
 		if f.dependencies.Cmk != nil {
 			return nil
 		}
+		execConfig, err := decoder.ParseCloudStackCredsFromEnv()
+		if err != nil {
+			return fmt.Errorf("parsing CloudStack credentials: %v", err)
+		}
 
-		execConfig, err := decoder.ParseCloudStackSecret()
+		f.dependencies.Cmk, err = f.executablesConfig.builder.BuildCmkExecutable(f.dependencies.Writer, execConfig)
 		if err != nil {
 			return fmt.Errorf("building cmk executable: %v", err)
 		}
+		f.dependencies.closers = append(f.dependencies.closers, f.dependencies.Cmk)
+
+		return nil
+	})
 
-		f.dependencies.Cmk = f.executablesConfig.builder.BuildCmkExecutable(f.dependencies.Writer, execConfig.Profiles)
+	return f
+}
+
+// WithCmkFromSecrets initializes an empty Cmk object to make it available in the controller, which will set the profiles from K8s secrets.
+func (f *Factory) WithCmkFromSecrets() *Factory {
```

Comment:
How is the controller going to set the config in the cmk executable?
Would that happen on every reconciliation request? That wouldn't be thread safe

If the cmk executable is designed to only operate on one set of profiles, then I think we should build a new cmk executable on every reconciliation request. In order to do that, we can have a `CmkFactory` that implements a `CloudstackClientRegistry`. From the controller perspective it would be something like `cloudstackClientRegistry.Get(cloudStackExecConfigFromSecret)`. And such factory would just know how to create the cmk executable. Which means it probably needs the executable builder and the writer.